Update to music-metadata from version 7.12.3 to 11.3.0 by switching to Node.js-aware ES module resolution

Borewit opened the pull request:
- Alternative to PR #1352, avoiding utilizing [load-esm](https://github.com/Borewit/load-esm), but tweak the TypeScript compiler so that it actually does preserve the dynamic import.
- Avoid to convert file to a stream, which prevents random access to files, which can be much slower in some cases
- Updates music-metadata from version 7.12.3 to 11.2.3

jlucaso1 commented on src/Utils/messages-media.ts:
```diff
@@ -4,7 +4,7 @@ import { exec } from 'child_process'
 import { once } from 'events'
 import { createReadStream, createWriteStream, promises as fs, WriteStream } from 'fs'
-import type { IAudioMetadata } from 'music-metadata'
+import type { IAudioMetadata } from 'music-metadata' with { 'resolution-mode' : 'require' }
```
this is an import type only. Its necessary to use resolution-mode?

Borewit replied:
Without it:

```
> baileys@6.7.16 build:tsc
> tsc

src/Utils/messages-media.ts:7:37 - error TS1541: Type-only import of an ECMAScript module from a CommonJS module must have a 'resolution-mode' attribute.
  To convert this file to an ECMAScript module, change its file extension to '.mts', or add the field `"type": "module"` to ...
```
but I think it better to use `resolution-mode: import`, given it is from an ESM module

jlucaso1 commented on src/Utils/messages-media.ts:
```diff
@@ -186,19 +186,18 @@ export const mediaMessageSHA256B64 = (message: WAMessageContent) => {
 }
 
 export async function getAudioDuration(buffer: Buffer | string | Readable) {
+	// load-esm is used to avoid the TypeScript compiler to convert the dynamic import to require
```
remove this comment from the another pr

AstroX11 replied:
Huh?

jlucaso1 commented:
I Liked this approach. Any regression in the generated js files?

AstroX11 reviewed (approved):
Thanks

purpshell reviewed (changes requested):
im not sure about the proposed changes to the tsconfig. I'm sure there are some environments where this breaks. 

Till I properly setup tsconfig to withhold these kinda imports, we can't really try to "modernize" the approach.

For the time being, "if it ain't broke, don't fix it". Also, your PR title is misleading

Borewit commented:
Thanks for the review, I understand the concern about the tsconfig change.

> I'm not sure about the proposed changes to the tsconfig. I'm sure there are some environments where this breaks.

Fair point, changing the module resolution strategy can introduce side effects, and I can’t guarantee it won’t impact tooling or other parts of the project.

> Till I properly setup tsconfig to withhold these kinda imports, we can't really try to 'modernize' the approach.

I don’t fully agree. As demonstrated in [PR #1352](https://github.com/WhiskeySockets/Baileys/pull/1352), it’s entirely possible to upgrade [music-metadata](https://github.com/Borewit/music-metadata) without changing the tsconfig.

> If it ain't broke, don't fix it.

That’s reasonable — but music-metadata [7.12.3](https://github.com/Borewit/music-metadata/releases/tag/v7.12.3) has had plenty of fixes since then, many addressing issues that only show up with specific metadata structures. Whether that’s worth addressing now is up to you.

>Your PR title is misleading.

I believe the title describes was this PR is about, there is certainly nothing misleading in that.

purpshell commented:
> Thanks for the review, I understand the concern about the tsconfig change.
> 
> > I'm not sure about the proposed changes to the tsconfig. I'm sure there are some environments where this breaks.
> 
> Fair point, changing the module resolution strategy can introduce side effects, and I can’t guarantee it won’t impact tooling or other parts of the project.
> 
> > Till I properly setup tsconfig to withhold these kinda imports, we can't really try to 'modernize' the approach.
> 
> I don’t fully agree. As demonstrated in [PR #1352](https://github.com/WhiskeySockets/Baileys/pull/1352), it’s entirely possible to upgrade [music-metadata](https://github.com/Borewit/music-metadata) without changing the tsconfig.
> 
> > If it ain't broke, don't fix it.
> 
> That’s reasonable — but music-metadata [7.12.3](https://github.com/Borewit/music-metadata/releases/tag/v7.12.3) has had plenty of fixes since then, many addressing issues that only show up with specific metadata structures. Whether that’s worth addressing now is up to you.
> 
> > Your PR title is misleading.
> 
> I believe the title describes was this PR is about, there is certainly nothing misleading in that.

Yeah but I don't want to introduce things like loadEsm or the likes. For the time being, it is sufficient to use the outdated version as is. I'll keep this pull request open until we solve the ESM issue with Baileys (current proposed solution is to completely get rid of CommonJS). Thank you for your contributions, I appreciate that the maintainer of the library is helping its users as well (and not just maintaining the project which is pretty useful to us!)

purpshell commented:
> > Yeah but I don't want to introduce things like loadEsm or the likes. For the time being, it is sufficient to use the outdated version as is.
> 
> I understand it may not look appealing to load an ESM module via another module, but please be aware this is a super tiny module (nearly a one liner):
> 
> ```js
> // load-esm/index.js:
> 
> module.exports = {
>     loadEsm: (module) => import(module)
> };
> ```
> 
> The only thing is does, is to keep the `await import` outside the scope of the TypeScript compiler, to prevent it from converting `await import` to `require`.
> 
> Download counter of [load-esm](https://github.com/Borewit/load-esm) is on 5 million downloads a month, used in at least 25 thousands projects, not a single issue reported, so far 🤞🏻 .
> 
> You can also hack it it without `load-esm`, using an eval expression (not recommended), or using a local `.js` file.
> 
> More here on alternatives: https://stackoverflow.com/a/79265806/28701779
> 
> > I appreciate that the maintainer of the library is helping its users as well (and not just maintaining the project which is pretty useful to us!)
> 
> ❤️

I am in the process of turning Baileys into ESM. The majority by a landslide wants it to be an ESM module. Thank you for your suggestions!

Borewit commented:
> I am in the process of turning Baileys into ESM. The majority by a landslide wants it to be an ESM module. Thank you for your suggestions!

I can only encourage migrating to ESM. For your users, who are stuck with a TypeScript/CommonJS, after your migration, you may refer to the solutions provided.

AstroX11 commented:
@Borewit Hi, please could you update your PR, Baileys is now ESM

purpshell commented:
> @Borewit Hi, please could you update your PR, Baileys is now ESM
I'll work on it personally. 


@Borewit I have applied the changes necessary to move Baileys to ESM and it is ready now. Thank you for the support.
